### ISSUE TITLE
XLA single-TPU code runs as slowly as CPU on my NLP code

### ISSUE THREAD

**Issue by githubrandomuser2017** [pytorch/xla#2673]
## 🐛 Bug

I am trying to train a natural language processing model (using the HuggingFace Transformers library) on Google Colab. I followed the XLA instructions (found [here](https://medium.com/pytorch/get-started-with-pytorch-cloud-tpus-and-colab-a24757b8f7fc) and [here](https://pytorch.org/xla/release/1.7/index.html#running-models-on-xla-devices)). Specifically, I did the following: 
- created a `device` with `xm.xla_device()`
- moved my model and tensors to that device with `.to(device)`

However, the training with XLA and single-TPU usage is very **slow**, likely as slow as CPU.

## To Reproduce

I created a [Google Colab notebook](https://colab.research.google.com/drive/1up7f3CvOLRhXYXEBlcA_oZxcaz9MxipW?usp=sharing) showing my code and how slow it is to train.

## Expected behavior

I would expect some kind of training time speedup using a single TPU core, not the running time of CPU training.

## Environment

 - Reproducible on XLA backend [CPU/TPU]: TPU
 - torch_xla version: 1.7

**Comment by jysohn23** [comment]
Could you try using the stuff we've built into HF's trainer? Check out this documentation: https://github.com/huggingface/transformers/tree/master/examples#running-on-tpus

We run daily tests on these and keep track of performance.

**Comment by githubrandomuser2017** [comment]
I don't user the HuggingFace `Trainer` API. I implement my own PyTorch training loop for this simple single-TPU training task. I was expecting that I only need to set `device` correctly, which I think I've done.

**Comment by jysohn23** [comment]
We recommend using the `trainer.py` since it already has a lot of the optimizations implemented there as mentioned in https://github.com/pytorch/xla/blob/master/API_GUIDE.md

So it looks like you're missing a couple things which you can find in the above linked API_GUIDE.md:
* Use `pl.MpDeviceLoader` which overlaps computation, input feeding, and takes care of marking steps.
* Do not call `.item()` like you're doing in `multiclass_accuracy` method at every step. This hurts performance as described in https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#known-performance-caveats under "Certain operations don't have native translations to XLA.".

**Comment by githubrandomuser2017** [comment]
What is `trainer.py`? Is that part of XLA or part of HuggingFace Transformers?

I see that `pl.MpDeviceLoader()` is used for multiprocessing. At this time, I'm just trying to run my code on one TPU. Is that function still relevant to me?

**Comment by taylanbil** [comment]
yes, it is highly recommended to use that in all settings, 1 or 8 or more cores.

**Comment by jysohn23** [comment]
https://github.com/huggingface/transformers/blob/master/src/transformers/trainer.py

**Comment by githubrandomuser2017** [comment]
Are there any successful examples of XLA working with the HuggingFace Transformers library using bare PyTorch and not a wrapper API like `Trainer()`? I followed all instructions I found on the help pages listed above, including removing any calls to `.item()` (and replacing them with `.detach()`), but my code still takes a long time just to finish one epoch. 

I guess my only recourse to use TPUs is to port my code to use `Trainer()`.

**Comment by jysohn23** [comment]
The HuggingFace `Trainer` module does pretty much what's described in our [API_GUIDE](https://github.com/pytorch/xla/blob/master/API_GUIDE.md) and [TROUBLESHOOTING](https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md) docs to enable running on TPUs. In addition to the `trainer.py` script, I'd had to update different modules that it depended on as well. A quick [search on `torch_xla`](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/huggingface/transformers%24+torch_xla&patternType=literal). But yeah there shouldn't be anything magical about the `Trainer` module, it just already has the TPU support added so it's much easier to use since you don't need to duplicate the work.
